cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredAgentScheduler.java

```diff
@@ -43,240 +39,260 @@
 
 @SuppressFBWarnings
 public class ClusteredAgentScheduler extends CatsModuleAware implements AgentScheduler<AgentLock>, Runnable {
-    private static enum Status {
-        SUCCESS,
-        FAILURE
+  private static enum Status {
+    SUCCESS,
+    FAILURE
+  }
+
+  private static final Logger logger = LoggerFactory.getLogger(ClusteredAgentScheduler.class);
+
+  private final RedisClientDelegate redisClientDelegate;
+  private final NodeIdentity nodeIdentity;
+  private final AgentIntervalProvider intervalProvider;
+  private final ExecutorService agentExecutionPool;
+  private final Pattern enabledAgentPattern;
+
+  private final Map<String, AgentExecutionAction> agents = new ConcurrentHashMap<>();
+  private final Map<String, NextAttempt> activeAgents = new ConcurrentHashMap<>();
+  private final NodeStatusProvider nodeStatusProvider;
+  private Integer maxConcurrentAgents;
+
+  public ClusteredAgentScheduler(RedisClientDelegate redisClientDelegate,
+                                 NodeIdentity nodeIdentity,
+                                 AgentIntervalProvider intervalProvider,
+                                 NodeStatusProvider nodeStatusProvider,
+                                 String enabledAgentPattern,
+                                 Integer maxConcurrentAgents,
+                                 Integer agentLockAcquisitionIntervalSeconds) {
+    this(
+      redisClientDelegate,
+      nodeIdentity,
+      intervalProvider,
+      nodeStatusProvider,
+      Executors.newSingleThreadScheduledExecutor(new NamedThreadFactory(ClusteredAgentScheduler.class.getSimpleName())),
+      Executors.newCachedThreadPool(new NamedThreadFactory(AgentExecutionAction.class.getSimpleName())),
+      enabledAgentPattern,
+      maxConcurrentAgents,
+      agentLockAcquisitionIntervalSeconds
+    );
+  }
+
+  public ClusteredAgentScheduler(RedisClientDelegate redisClientDelegate,
+                                 NodeIdentity nodeIdentity,
+                                 AgentIntervalProvider intervalProvider,
+                                 NodeStatusProvider nodeStatusProvider,
+                                 ScheduledExecutorService lockPollingScheduler,
+                                 ExecutorService agentExecutionPool,
+                                 String enabledAgentPattern,
+                                 Integer maxConcurrentAgents,
+                                 Integer agentLockAcquisitionIntervalSeconds) {
+    this.redisClientDelegate = redisClientDelegate;
+    this.nodeIdentity = nodeIdentity;
+    this.intervalProvider = intervalProvider;
+    this.nodeStatusProvider = nodeStatusProvider;
+    this.agentExecutionPool = agentExecutionPool;
+    this.enabledAgentPattern = Pattern.compile(enabledAgentPattern);
+    this.maxConcurrentAgents = maxConcurrentAgents == null ? Integer.MAX_VALUE : maxConcurrentAgents;
```

Comment:
Do we really want to have `Integer.MAX_VALUE` as the default? If we're going to have a default, I don't think it should be a value that we've demonstrated as being pathological.

Reply:
The goal is to keep the behavior the same. `Integer.MAX_VALUE` is definitely too high ... but it does keep the behavior the same. Thoughts on changing the behavior to something more sensible? 100?

Reply:
➕

Reply:
That works for me.